### PR TITLE
Changed example to use non-HTTPS site to avoid 301 errors

### PR DESCRIPTION
**Pull request by geky**
Adopted ipify.org (thanks @bridadan). This avoids the redirect for HTTPS.

```
IP address: 10.118.14.45
Netmask: 255.255.252.0
Gateway: 10.118.12.1
sent 39 [GET / HTTP/1.1]
recv 173 [HTTP/1.1 200 OK]
External IP address: 217.140.111.135
Done
```

Should take care of https://github.com/ARMmbed/mbed-os-example-sockets/issues/21
cc @maclobdell, @bridadan

### CODE REVIEW

File: main.cpp [hunk 1]
```diff
@@ -12,25 +12,34 @@ int main() {
...
+    printf("IP address: %s\n", ip ? ip : "None");
+    printf("Netmask: %s\n", netmask ? netmask : "None");
+    printf("Gateway: %s\n", gateway ? gateway : "None");
 
     // Open a socket on the network interface, and create a TCP connection to mbed.org
     TCPSocket socket;
     socket.open(&net);
-    socket.connect("developer.mbed.org", 80);
+    socket.connect("api.ipify.org", 80);
+    char *buffer = new char[256];
```
Comment (bridadan) [inline main.cpp:25]: Any reason this isn't on the stack?

**Review by bridadan — APPROVED** [review]

File: main.cpp (hunk 1 above)
Comment (geky) [reply main.cpp:25]: It was previously, but we probably shouldn't be encouraging stack allocations for buffers. Stack allocations work fine for this buffer since it's _relatively_ small, but if a user copied this code and tried to expand the buffer for a heavier protocol, they may quickly find themselves with stack overflows.

Comment (bridadan) [reply main.cpp:25]: > but we probably shouldn't be encouraging stack allocations for buffers. 

Forgive my ignorance here. Is it not common practice to put buffers on stacks? Or I guess since it has the chance of buffer overruns it's not so bueno?

Comment (geky) [reply main.cpp:25]: The best place to allocate is in global memory, since the compiler can statically check for an out-of-memory (oom) situation. Both heap and stack can be overflowed at runtime, but heap allocations are easier to reason about and track. When evaluating heap size, the first step is finding where heap allocations occur. But when evaluating stack size, you usually think in terms of functions calls, not taking into account the size of the objects allocated in the functions.

For most objects allocated on the stack, the overhead is comparable to a reasonable stack frame (~64bytes). But once you introduce buffers, the stack frame can quickly grow (in this example ~256 bytes for one buffer). The user needs to consider both what functions are called while the buffer is on the stack, and where the function is called from to make sure the stack is large enough. This isn't something most users think about when writing application code.

In the end, both stack and heap allocations end up out of RAM, stack is just less flexible and harder to track.

Comment (bridadan) [reply main.cpp:25]: OK, thanks for thorough explanation. I was mostly just curious (since I already approved the changes :smile:)

Comment (geky) [reply main.cpp:25]: No problem 😀

Comment (kjbracey) [reply main.cpp:25]: If we weren't using this RTOS, I'd be quite inclined to use the stack. I probably would have done in mbed OS 3. (And I think it's more predictable+reliable than the heap - the heap can fill up or become fragmented while running, the stack can't.)

But in this RTOS environment, stack usage can be particularly problematic. You need to allocate enough RAM for the peak usage of every thread. 

    Stack RAM = Sum<1..N> [MaxStackUsageForThread(N)]

You need to do everything possible to make each thread's stack usage be as small as possible to try to keep that sum down. It gets out of hand incredibly fast. Ideally, do not create any threads. Increasing the stack usage of any thread increases total RAM requirement.

If you avoid using the RTOS, and use event queue type mechanisms on single thread, then stack RAM would be more like

    Stack RAM = Max<1..N> [MaxStackUsageForEvent(N)]

Which is a /much/ smaller number, and increasing the stack usage of any one event is likely to actually just use already-paid-for stack, unless it becomes the biggest user.

In summary - event loop stacks or the heap are widely-shared resources - by using them for a temporary, you're likely to not increase total system RAM requirement. Thread stacks are less shared, and using them is more likely to increase total system RAM requirement - avoid them.

And to give some sense of scale - RTX thinks a sensible default stack size for a thread is 512 bytes. Not much room for 256-byte buffers on there.
